### PR TITLE
More stage matches

### PR DESCRIPTION
**Pull request by JaceCear**

### CODE REVIEW

File: src/game/stage/collect_ring_effect.c [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+#include "game/game.h"
+#include "game/stage/rings_manager.h"
+
+#include "constants/animations.h"
+
+typedef struct {
+    Sprite s;
+    u8 unk30;
+    u8 unk31;
+} Struct_RingEffect;
```
Comment (freshollie) [inline src/game/stage/collect_ring_effect.c:13]: Any reason to name `Struct_`?

File: src/game/stage/item_box.c [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+
+void CreateEntity_ItemBox(MapEntity *me, u16 spriteRegionX, u16 spriteRegionY,
+                          u8 spriteY)
+{
+    struct Task *t;
+    Entity_ItemBox *itembox;
+    Sprite *s;
+    if (gGameMode == GAME_MODE_TIME_ATTACK || gGameMode == GAME_MODE_BOSS_TIME_ATTACK) {
+        if (me->index == 0) {
+            goto CreateEntity_Itembox_defer;
```
Comment (freshollie) [inline src/game/stage/item_box.c:71]: Interesting. Did you try this with a return?

File: src/game/stage/item_box.c [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+            newRingCount = *rings + 5;
+            goto ApplyItemboxEffect_case7;
+        } break;
+
+        case 7: {
+            rings = &gRingCount;
+            oldRingCount = *rings;
+            newRingCount = *rings + 10;
+
+        ApplyItemboxEffect_case7:
```
[the remark this answers is not in the document]
Comment (freshollie) [reply src/game/stage/item_box.c:232]: Could be real though :/ good job!

File: src/game/stage/item_box.c [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+            // Find the player that's closest to you
+            for (playerId = 0; playerId < MULTI_SIO_PLAYERS_MAX
+                 && gMultiplayerPlayerTasks[playerId] != NULL;
+                 playerId++) {
+                struct MultiplayerPlayer *mpp;
+                s32 boxToPlayerX, boxToPlayerY;
+                s32 boxToPlayerMagnitude;
+
+                // Don't look for your own ID
+                u32 sioId = (*(vu32 *)&REG_SIOCNT << 26) >> 30;
```
Comment (freshollie) [inline src/game/stage/item_box.c:273]: There is a macro already for this `(SIO_MULTI_CNT)->id;`

File: src/game/stage/spot_light.c [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                    ut->unkB++;
+
+                ut = TaskGetStructPtr(spotLight->t1);
+
+                if (ut->unkB < 16)
+                    ut->unkB++;
+            } else {
+                goto goto_labl;
+            }
+        } else if (!(gDispCnt & DISPCNT_BG0_ON)) {
```
Comment (freshollie) [inline src/game/stage/spot_light.c:80]: If you add `&& (gDispCnt & DISPCNT_BG0_ON)` to the top level if, you may find that the match for the else go to is solved

Comment (freshollie) [reply src/game/stage/spot_light.c:80]: As the compiler will skip the `!(gDispCnt & DISPCNT_BG0_ON)` condition that it's already checked

File: src/game/time_attack/results.c [hunk 6]
```diff
@@ -50,7 +50,7 @@ u32 CreateTimeAttackResultsCutScene(u32 finishTime)
     u8 i;
     u8 isBossLevel;
     u8 level;
-    gLoadedSaveGame->score += gRingCount;
+    gLoadedSaveGame->score += (s16)gRingCount;
```
Comment (freshollie) [inline src/game/time_attack/results.c:53]: Makes sense (well doesn't make sense but ringCount as unsigned global is more obvious

File: src/game/stage/entities_manager.c [hunk 7]
```diff
@@ -1054,10 +1056,13 @@ void sub_8009530(s16 a, s16 b)
...
 
-    sub_8009E00(a, b);
+    CreateEnemyDefeatScore(x, y);
 
-    if (gPlayer.unk84 < 4) {
-        gPlayer.unk84++;
+    // NOTE: This should be (ARRAY_COUNT(enemyDefeatScores) - 1)
+    //       But padding makes it (6-1) instead of (5-1),
+    //       hence the macro.
+    if (gPlayer.defeatScoreIndex < (NUN_ENEMY_DEFEAT_SCORES - 1)) {
```
[the remark this answers is not in the document]
Comment (JaceCear) [reply src/game/stage/entities_manager.c:1064]: Nope.
I didn't mean manual padding, but padding added by GCC.
No idea why it's happening in this specific case.

File: src/game/stage/collect_ring_effect.c (hunk 1 above)
Comment (JaceCear) [reply src/game/stage/collect_ring_effect.c:13]: No special reason, I've just started doing it at some point. Could change that if you wanna.

File: src/game/stage/item_box.c (hunk 2 above)
Comment (JaceCear) [reply src/game/stage/item_box.c:71]: Yeah I did, but initially it didn't work.

But I tried again and it does, thanks!

Comment (JaceCear) [reply src/game/stage/item_box.c:71]: Though I feel like that the goto is "cleaner" than setting the ME init at the top and the bottom, idk.

File: src/game/stage/item_box.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (JaceCear) [reply src/game/stage/item_box.c:232]: Same here, I tried it in the beginning and it didn't work, but now it does.
And again I feel like the gotos make it more readable than copy-pasting the code in all 3 cases.
Might make a macro out of it.

File: src/game/stage/item_box.c (hunk 4 above)
Comment (JaceCear) [reply src/game/stage/item_box.c:273]: Awesome!

File: src/game/stage/spot_light.c (hunk 5 above)
Comment (JaceCear) [reply src/game/stage/spot_light.c:80]: I might be doing something wrong, but that doesn't work for me.

File: src/game/time_attack/results.c (hunk 6 above)
Comment (JaceCear) [reply src/game/time_attack/results.c:53]: Yeah, that's what I was thinking!
And it's better to cast the odd case than the common case!

File: src/game/stage/entities_manager.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (freshollie) [reply src/game/stage/entities_manager.c:1064]: Oh weird!

File: src/game/stage/item_box.c (hunk 2 above)
Comment (freshollie) [reply src/game/stage/item_box.c]: Nah, I think it's best to write what the developers probably wrote. So far this seems to be a pattern

File: src/game/stage/item_box.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (freshollie) [reply src/game/stage/item_box.c]: Yep, make a macro as that's probably what they wrote :)

File: src/game/stage/spot_light.c (hunk 5 above)
Comment (freshollie) [reply src/game/stage/spot_light.c:80]: I will have a go

Comment (freshollie) [reply src/game/stage/spot_light.c:78]: Fixed

**Review by freshollie — APPROVED** [review]

**Pull request merged**